sycl/doc/extensions/GroupAlgorithms/SYCL_INTEL_group_algorithms.asciidoc

```diff
@@ -0,0 +1,268 @@
+= SYCL_INTEL_group_algorithms
+:source-highlighter: coderay
+:coderay-linenums-mode: table
+
+// This section needs to be after the document title.
+:doctype: book
+:toc2:
+:toc: left
+:encoding: utf-8
+:lang: en
+
+:blank: pass:[ +]
+
+// Set the default source code type in this document to C++,
+// for syntax highlighting purposes.  This is needed because
+// docbook uses c++ and html5 uses cpp.
+:language: {basebackend@docbook:c++:cpp}
+
+== Introduction
+IMPORTANT: This specification is a draft.
+
+NOTE: Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by permission by Khronos.
+
+NOTE: This document is better viewed when rendered as html with asciidoctor.  GitHub does not render image icons.
+
+This document describes an extension which introduces a library of group functions, including common parallel algorithms such as reductions and scans.
+
+== Name Strings
+
++SYCL_INTEL_group_algorithms+
+
+== Notice
+
+Copyright (c) 2020 Intel Corporation.  All rights reserved.
+
+== Status
+
+Working Draft
+
+This is a preview extension specification, intended to provide early access to a feature for review and community feedback. When the feature matures, this specification may be released as a formal extension.
+
+Because the interfaces defined by this specification are not final and are subject to change they are not intended to be used by shipping software products.
+
+== Version
+
+Built On: {docdate} +
+Revision: 1
+
+== Contact
+John Pennycook, Intel (john 'dot' pennycook 'at' intel 'dot' com)
+
+== Dependencies
+
+This extension is written against the SYCL 1.2.1 specification, Revision v1.2.1-6.
+
+== Overview
+
+This extension introduces a library of group algorithms, providing a simple way for developers to apply common parallel algorithms across data held by different work-items in the same group, or to use the work-items of a group to apply common parallel algorithms across data stored in memory.
+
+The extension introduces the following functions:
+
+- +any_of+
+- +all_of+
+- +none_of+
+- +broadcast+
```

Comment:
I agree that `fill` is the closest existing algorithm to the behavior of `broadcast`.  But after reviewing the others, I noticed there's a pattern they all seem to follow that wouldn't be the same here.

If we generalize a little, the algorithms in C++17 look like one of the two declarations below:
```c++
ReturnT algorithm(InputIt first, InputIt last, OtherArgs...);
void algorithm(InputIt first, InputIt last, OutputIt result, OtherArgs...);
```

When an algorithm of the first form becomes a group algorithm: the input range becomes a variable held by each work-item; the return value is broadcast to all work-items in the group; and the other arguments are required to be uniform.

When an algorithm of the second form becomes a group algorithm: the input range becomes a variable held by each work-item; the result is distributed across the work-items in the group; and the other arguments are required to be uniform.

If we are able to follow this pattern consistently, it will be easier for us to convert other algorithms in future and easier for developers to reason about expected behavior.  If we have a case that doesn't fit, maybe that means we need a new range variant.

If we apply the thinking above to `std::fill`, we'd get:
```c++
T fill(const T& value);
```
...which makes sense, but isn't the behavior of `broadcast`.

What we'd want, I think, is for the range variant to look something like this:
```c++
void broadcast(Ptr first, Ptr last, ptrdiff_t offset = 0);
// sets every element of [first, last) to *first + offset -- equivalent to std::fill(first, last, first[offset])
```

What do you think?  Does it make sense to define something like this function for symmetry?  If so, would you still recommend changing its name from `broadcast` to something else?

Reply:
@rolandschulz and I discussed this offline and reached the conclusion that we do want to follow this pattern for consistency, and so we shouldn't rename `broadcast` to `fill`.